File: examples3d/debug_cylinder3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

// This shows a bug when a cylinder is in contact with a very large
// but very thin cuboid. In this case the EPA returns an incorrect
// contact normal, resulting in the cylinder falling through the floor.
pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 100.1;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height, 0.0));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    let (_handle, _) = world.insert(rigid_body, collider);

    /*
     * Create the cubes
     */
    let num = 1;
    let rad = 1.0;

    let shiftx = rad * 2.0 + rad;
    let shifty = rad * 2.0 + rad;
    let shiftz = rad * 2.0 + rad;
    let centerx = shiftx * (num / 2) as f32;
    let centery = shifty / 2.0;
    let centerz = shiftz * (num / 2) as f32;

    let offset = -(num as f32) * (rad * 2.0 + rad) * 0.5;

    let x = -centerx + offset;
    let y = centery + 3.0;
    let z = -centerz + offset;

    // Build the rigid body.
    let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(x, y, z));
    let collider = ColliderBuilder::cylinder(rad, rad);
    let (_handle, _) = world.insert(rigid_body, collider);

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(100.0, 100.0, 100.0), Vec3::ZERO);
}

File: python/rapier-testbed/rapier_testbed/examples3/debug_cylinder3.py
"""Port of ``examples3d/debug_cylinder3.rs``.

A single cylinder dropping onto a very thin cuboid floor — repros an
older EPA bug where the cylinder fell through the floor.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Debug"
NAME = "Cylinder"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 100.1
    ground_height = 0.1
    ground_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size),
        ground_handle,
        bodies,
    )

    rad = 1.0
    centery = (rad * 2.0 + rad) / 2.0
    offset = -(1.0) * (rad * 2.0 + rad) * 0.5
    x = 0.0 + offset
    y = centery + 3.0
    z = 0.0 + offset

    body = rp.RigidBody.dynamic().translation((x, y, z))
    handle = bodies.insert(body)
    colliders.insert_with_parent(rp.Collider.cylinder(rad, rad), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
